model-engine/model_engine_server/infra/repositories/live_tokenizer_repository.py

```diff
@@ -0,0 +1,149 @@
+        "gpt4all-j": ModelInfo("nomic-ai/gpt4all-j", None),
+        "dolly-v2-12b": ModelInfo("databricks/dolly-v2-12b", None),
+        "stablelm-tuned-7b": ModelInfo("StabilityAI/stablelm-tuned-alpha-7b", None),
+        "vicuna-13b": ModelInfo("eachadea/vicuna-13b-1.1", None),
+    }
+
+
+def get_supported_models_info() -> Dict[str, ModelInfo]:
+    try:
+        from plugins.live_tokenizer_repository import (
+            get_supported_models_info as get_custom_supported_models_info,
+        )
+
+        return get_custom_supported_models_info()
+    except ModuleNotFoundError:
+        return get_default_supported_models_info()
+
+
+SUPPORTED_MODELS_INFO = get_supported_models_info()
+
+
+def get_models_s3_uri(*args, **kwargs) -> str:
+    try:
+        from plugins.live_tokenizer_repository import get_models_s3_uri as get_custom_models_s3_uri
+
+        return get_custom_models_s3_uri(*args, **kwargs)
+    except ModuleNotFoundError:
+        raise NotImplementedError
+
+
+def get_models_local_dir_path(model_name: str) -> str:
+    """
+    Get the local directory path for a given model.
+    """
+    return f"{TOKENIZER_TARGET_DIR}/{model_name}"
+
+
+class LiveTokenizerRepository(TokenizerRepository):
+    def __init__(self, llm_artifact_gateway: LLMArtifactGateway):
+        self.llm_artifact_gateway = llm_artifact_gateway
+
+    def _load_tokenizer_from_s3(self, model_name: str, s3_prefix: Optional[str]) -> Optional[str]:
+        """
+        Download tokenizer files from S3 to the local filesystem.
+        """
+        if not s3_prefix:
+            return None
+
+        model_tokenizer_dir = get_models_local_dir_path(model_name)
+
+        for file in TOKENIZER_FILES_REQUIRED:
+            s3_path = get_models_s3_uri(s3_prefix, file)
+            target_path = os.path.join(model_tokenizer_dir, file)
+            self.llm_artifact_gateway.download_files(s3_path, target_path)
+
+        for file in TOKENIZER_FILES_OPTIONAL:
+            s3_path = get_models_s3_uri(s3_prefix, file)
+            target_path = os.path.join(model_tokenizer_dir, file)
+            try:
+                self.llm_artifact_gateway.download_files(s3_path, target_path)
+            except Exception:
+                pass
+
+        return model_tokenizer_dir
+
+    @lru_cache(maxsize=32)
+    def load_tokenizer(self, model_name: str) -> AutoTokenizer:
+        model_info = SUPPORTED_MODELS_INFO[model_name]
+
+        model_location = None
+        try:
+            if not model_info.hf_repo:
+                raise RepositoryNotFoundError("No HF repo specified for model.")
+            list_repo_refs(model_info.hf_repo)  # check if model exists in Hugging Face Hub
+            model_location = model_info.hf_repo
+            # AutoTokenizer handles file downloads for HF repos
+        except RepositoryNotFoundError:
+            model_location = self._load_tokenizer_from_s3(model_name, model_info.s3_repo)
```

Comment:
What's the behavior if this fails? Seems like the API layer would catch and throw a 500?

Reply:
Oh nevermind I'm guess there's a catch-all `None` response based on the `if not model_location` call below.